--- rhino3dm.js	1985-10-26 09:15:00.000000000 +0100
+++ rhino3dm.module.js	1985-10-26 09:15:00.000000000 +0100
@@ -1,7 +1,7 @@
 
 var rhino3dm = (function() {
-  var _scriptDir = typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined;
-  if (typeof __filename !== 'undefined') _scriptDir = _scriptDir || __filename;
+  var _scriptDir = import.meta.url;
+  
   return (
 function(rhino3dm) {
   rhino3dm = rhino3dm || {};
@@ -13,9 +13,4 @@
 }
 );
 })();
-if (typeof exports === 'object' && typeof module === 'object')
-  module.exports = rhino3dm;
-else if (typeof define === 'function' && define['amd'])
-  define([], function() { return rhino3dm; });
-else if (typeof exports === 'object')
-  exports["rhino3dm"] = rhino3dm;
+export default rhino3dm;
\ No newline at end of file
